dask/array/core.py

```diff
@@ -893,20 +897,61 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
         except AttributeError:
             dsk = {}
 
-        update = insert_to_ooc(src, tgt, lock=lock, region=reg)
-        keys.extend(update)
-
-        update.update(src.dask)
-        update.update(dsk)
-        updates.update(update)
+        each_store_dsk = insert_to_ooc(src, tgt, lock=lock, region=reg, keep=keep)
+
+        if keep:
+            load_names.append('load-%s' % src.name)
+            load_dsks.append(retrieve_from_ooc(each_store_dsk))
+
+        for each_store_key in each_store_dsk:
+            store_dlyds.append(
+                Delayed(
+                    each_store_key,
+                    sharedict.merge(
+                        (
+                            each_store_key,
+                            {each_store_key: each_store_dsk[each_store_key]}
+                        ),
+                        dsk,
+                        src.dask
+                    )
+                )
+            )
 
-    name = 'store-' + tokenize(*keys)
-    dsk = sharedict.merge({name: keys}, updates)
-    result = Delayed(name, dsk)
-    if compute:
-        result.compute()
+    if keep:
+        if compute:
+            from ..base import persist
+            store_dlyds = persist(*store_dlyds)
```

Comment:
I think that we lose the dask.array optimizations here.  Being able to call `persist` on a dask.array here would be advantageous.  I probably gave bad advice in https://github.com/dask/dask/pull/3000 as well.  Will comment there shortly.

Reply:
If I recall correctly your previous solution involved a `map_blocks` call that assigned the value to the store and then returned that input downstream.  I apologize if I dissuaded you from this approach.  

Sorry in general for being short on time recently.

Reply:
No worries. It's the holidays after all. :)

Hmm...not sure that I follow the `persist` suggestion. While a Dask Array does go in as input, currently we are returning `out`, which could be...anything? Does `Delayed` not make sense here? How do we call `persist` on the `store` step without wrapping it in `Delayed`?

I don't think I was using `map_blocks` previously. Though I guess we could look at that option too. Is it ok for `map_block` to return an `object` array? Would it coerce `out` into something in memory (e.g. thinking of a persistent Zarr Array converted into a NumPy Array)?

Reply:
When we call compute/persist we first optimize the graph based on the type of the collection.  In the case of delayed objects there isn't much optimization to do.  However when the collection is a dask.array a ton of stuff gets run that fuses slicing, atop calls (eventually), and other special cases. By moving the task graph from an array to a delayed object without going through an optimization phase we've sidestepped calling optimizations on that graph that should have been called.

It is fine returning a delayed object, but we should call `__dask_optimize__` on the task graph before letting go of the array.

Reply:
Of course, that sounds totally reasonable. Happy to add a call to `__dask_optimize__` earlier on.

Reply:
Did this for `src` before calling `insert_to_ooc` in the `for`-loop. So will be reused for all the other cases. Not sure if that is what you had in mind though.